Hmmm... need to check why it's failing here

```
=== RUN   TestDockerSuite/TestBuildDockerignore
    docker_cli_build_test.go:2294: assertion failed: 
        Command:  d:\CI\PR-42684\1\binary\docker.exe build -t testbuilddockerignore .
        ExitCode: 1
        Error:    exit status 1
        Stdout:   Sending build context to Docker daemon  5.632kB

        Step 1/13 : FROM busybox
         ---> d5680261fb40
        Step 2/13 : ADD . /bla
        
        Stderr:   ADD failed: copy from c:\ or c:\windows is not allowed on windows
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerSuite/TestBuildDockerignore (0.08s)
```

--- a/builder/dockerfile/copy.go
+++ b/builder/dockerfile/copy.go
@@ -251,19 +251,6 @@ func (o *copier) calcCopyInfo(origPath string, allowWildcards bool) ([]copyInfo,
 	return newCopyInfos(newCopyInfoFromSource(o.source, origPath, hash)), nil
 }
 
-func containsWildcards(name string) bool {
-	isWindows := runtime.GOOS == "windows"
-	for i := 0; i < len(name); i++ {
-		ch := name[i]
-		if ch == '\\' && !isWindows {
-			i++
-		} else if ch == '*' || ch == '?' || ch == '[' {
-			return true
-		}
-	}
-	return false
-}
-
 func (o *copier) storeInPathCache(im *imageMount, path string, hash string) {
 	if im != nil {
 		o.pathCache.Store(im.ImageID()+path, hash)

--- a/builder/dockerfile/copy.go
+++ b/builder/dockerfile/copy.go
@@ -109,22 +109,15 @@ func copierFromDispatchRequest(req dispatchRequest, download sourceDownloader, i
 }
 
 func (o *copier) createCopyInstruction(sourcesAndDest instructions.SourcesAndDest, cmdName string) (copyInstruction, error) {
-	inst := copyInstruction{cmdName: cmdName}
-
-	// Work in platform-specific filepath semantics
-	// TODO: This OS switch for paths is NOT correct and should not be supported.
-	// Maintained for backwards compatibility
-	pathOS := runtime.GOOS
-	if o.platform != nil {
-		pathOS = o.platform.OS
+	inst := copyInstruction{
+		cmdName: cmdName,
+		dest:    filepath.FromSlash(sourcesAndDest.DestPath),
 	}
-	inst.dest = fromSlash(sourcesAndDest.DestPath, pathOS)
-	separator := string(separator(pathOS))
 	infos, err := o.getCopyInfosForSourcePaths(sourcesAndDest.SourcePaths, inst.dest)
 	if err != nil {
 		return inst, errors.Wrapf(err, "%s failed", cmdName)
 	}
-	if len(infos) > 1 && !strings.HasSuffix(inst.dest, separator) {
+	if len(infos) > 1 && !strings.HasSuffix(inst.dest, string(os.PathSeparator)) {
 		return inst, errors.Errorf("When using %s with more than one source file, the destination must be a directory and end with a /", cmdName)
 	}
 	inst.infos = infos

--- a/builder/dockerfile/copy.go
+++ b/builder/dockerfile/copy.go
@@ -191,6 +191,9 @@ func (o *copier) Cleanup() {
 // TODO: allowWildcards can probably be removed by refactoring this function further.
 func (o *copier) calcCopyInfo(origPath string, allowWildcards bool) ([]copyInfo, error) {
 	imageSource := o.imageSource
+	if err := validateCopySourcePath(imageSource, origPath); err != nil {
+		return nil, err
+	}
 
 	// TODO: do this when creating copier. Requires validateCopySourcePath
 	// (and other below) to be aware of the difference sources. Why is it only
@@ -215,20 +218,13 @@ func (o *copier) calcCopyInfo(origPath string, allowWildcards bool) ([]copyInfo,
 		return nil, errors.Errorf("missing build context")
 	}
 
-	root := o.source.Root()
-
-	if err := validateCopySourcePath(imageSource, origPath, root.OS()); err != nil {
-		return nil, err
-	}
-
-	// Work in source OS specific filepath semantics
-	// For LCOW, this is NOT the daemon OS.
-	origPath = root.FromSlash(origPath)
-	origPath = strings.TrimPrefix(origPath, string(root.Separator()))
-	origPath = strings.TrimPrefix(origPath, "."+string(root.Separator()))
+	// Work in daemon-specific OS filepath semantics
+	origPath = filepath.FromSlash(origPath)
+	origPath = strings.TrimPrefix(origPath, string(os.PathSeparator))
+	origPath = strings.TrimPrefix(origPath, "."+string(os.PathSeparator))
 
 	// Deal with wildcards
-	if allowWildcards && containsWildcards(origPath, root.OS()) {
+	if allowWildcards && containsWildcards(origPath) {
 		return o.copyWithWildcards(origPath)
 	}
 
@@ -262,8 +258,8 @@ func (o *copier) calcCopyInfo(origPath string, allowWildcards bool) ([]copyInfo,
 	return newCopyInfos(newCopyInfoFromSource(o.source, origPath, hash)), nil
 }
 
-func containsWildcards(name, platform string) bool {
-	isWindows := platform == "windows"
+func containsWildcards(name string) bool {
+	isWindows := runtime.GOOS == "windows"
 	for i := 0; i < len(name); i++ {
 		ch := name[i]
 		if ch == '\\' && !isWindows {
@@ -549,33 +545,23 @@ func copyDirectory(archiver Archiver, source, dest *copyEndpoint, identity *idto
 		return errors.Wrapf(err, "failed to copy directory")
 	}
 	if identity != nil {
-		// TODO: @gupta-ak. Investigate how LCOW permission mappings will work.
 		return fixPermissions(source.path, dest.path, *identity, !destExists)
 	}
 	return nil
 }
 
 func copyFile(archiver Archiver, source, dest *copyEndpoint, identity *idtools.Identity) error {
-	if runtime.GOOS == "windows" && dest.driver.OS() == "linux" {
-		// LCOW
-		if err := dest.driver.MkdirAll(dest.driver.Dir(dest.path), 0755); err != nil {
-			return errors.Wrapf(err, "failed to create new directory")
+	if identity == nil {
+		// Use system.MkdirAll here, which is a custom version of os.MkdirAll
+		// modified for use on Windows to handle volume GUID paths. These paths
+		// are of the form \\?\Volume{<GUID>}\<path>. An example would be:
+		// \\?\Volume{dae8d3ac-b9a1-11e9-88eb-e8554b2ba1db}\bin\busybox.exe
+		if err := system.MkdirAll(filepath.Dir(dest.path), 0755); err != nil {
+			return err
 		}
 	} else {
-		// Normal containers
-		if identity == nil {
-			// Use system.MkdirAll here, which is a custom version of os.MkdirAll
-			// modified for use on Windows to handle volume GUID paths. These paths
-			// are of the form \\?\Volume{<GUID>}\<path>. An example would be:
-			// \\?\Volume{dae8d3ac-b9a1-11e9-88eb-e8554b2ba1db}\bin\busybox.exe
-
-			if err := system.MkdirAll(filepath.Dir(dest.path), 0755); err != nil {
-				return err
-			}
-		} else {
-			if err := idtools.MkdirAllAndChownNew(filepath.Dir(dest.path), 0755, *identity); err != nil {
-				return errors.Wrapf(err, "failed to create new directory")
-			}
+		if err := idtools.MkdirAllAndChownNew(filepath.Dir(dest.path), 0755, *identity); err != nil {
+			return errors.Wrapf(err, "failed to create new directory")
 		}
 	}
 
@@ -583,7 +569,6 @@ func copyFile(archiver Archiver, source, dest *copyEndpoint, identity *idtools.I
 		return errors.Wrapf(err, "failed to copy file")
 	}
 	if identity != nil {
-		// TODO: @gupta-ak. Investigate how LCOW permission mappings will work.
 		return fixPermissions(source.path, dest.path, *identity, false)
 	}
 	return nil

--- a/builder/dockerfile/copy_unix.go
+++ b/builder/dockerfile/copy_unix.go
@@ -43,6 +43,6 @@ func fixPermissions(source, destination string, identity idtools.Identity, overr
 	})
 }
 
-func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
+func validateCopySourcePath(imageSource *imageMount, origPath string) error {
 	return nil
 }

--- a/builder/dockerfile/copy_unix.go
+++ b/builder/dockerfile/copy_unix.go
@@ -43,6 +43,18 @@ func fixPermissions(source, destination string, identity idtools.Identity, overr
 	})
 }
 
+func containsWildcards(name string) bool {
+	for i := 0; i < len(name); i++ {
+		ch := name[i]
+		if ch == '\\' {
+			i++
+		} else if ch == '*' || ch == '?' || ch == '[' {
+			return true
+		}
+	}
+	return false
+}
+
 func validateCopySourcePath(imageSource *imageMount, origPath string) error {
 	return nil
 }

--- a/builder/dockerfile/copy_windows.go
+++ b/builder/dockerfile/copy_windows.go
@@ -79,6 +79,16 @@ func fixPermissionsWindows(source, destination, SID string) error {
 	return windows.SetNamedSecurityInfo(destination, windows.SE_FILE_OBJECT, windows.OWNER_SECURITY_INFORMATION|windows.DACL_SECURITY_INFORMATION, sid, nil, dacl, nil)
 }
 
+func containsWildcards(name string) bool {
+	for i := 0; i < len(name); i++ {
+		ch := name[i]
+		if ch == '*' || ch == '?' || ch == '[' {
+			return true
+		}
+	}
+	return false
+}
+
 func validateCopySourcePath(imageSource *imageMount, origPath string) error {
 	if imageSource == nil {
 		return nil

--- a/builder/dockerfile/copy_windows.go
+++ b/builder/dockerfile/copy_windows.go
@@ -79,12 +79,10 @@ func fixPermissionsWindows(source, destination, SID string) error {
 	return windows.SetNamedSecurityInfo(destination, windows.SE_FILE_OBJECT, windows.OWNER_SECURITY_INFORMATION|windows.DACL_SECURITY_INFORMATION, sid, nil, dacl, nil)
 }
 
-func validateCopySourcePath(imageSource *imageMount, origPath, platform string) error {
-	// validate windows paths from other images + LCOW
-	if imageSource == nil || platform != "windows" {
+func validateCopySourcePath(imageSource *imageMount, origPath string) error {
+	if imageSource == nil {
 		return nil
 	}
-
 	origPath = filepath.FromSlash(origPath)
 	p := strings.ToLower(filepath.Clean(origPath))
 	if !filepath.IsAbs(p) {
@@ -92,13 +90,13 @@ func validateCopySourcePath(imageSource *imageMount, origPath, platform string) 
 			if p[len(p)-2:] == ":." { // case where clean returns weird c:. paths
 				p = p[:len(p)-1]
 			}
-			p += "\\"
+			p += `\`
 		} else {
-			p = filepath.Join("c:\\", p)
+			p = filepath.Join(`c:\`, p)
 		}
 	}
 	if _, ok := pathDenyList[p]; ok {
-		return errors.New("copy from c:\\ or c:\\windows is not allowed on windows")
+		return errors.New(`copy from c:\ or c:\windows is not allowed on windows`)
 	}
 	return nil
 }

--- a/builder/dockerfile/internals.go
+++ b/builder/dockerfile/internals.go
@@ -235,8 +235,8 @@ func createDestInfo(workingDir string, inst copyInstruction, rwLayer builder.RWL
 // normalizeDest normalises the destination of a COPY/ADD command in a
 // platform semantically consistent way.
 func normalizeDest(workingDir, requested string, platform string) (string, error) {
-	dest := fromSlash(requested, platform)
-	endsInSlash := strings.HasSuffix(dest, string(separator(platform)))
+	dest := filepath.FromSlash(requested)
+	endsInSlash := strings.HasSuffix(dest, string(os.PathSeparator))
 
 	if platform != "windows" {
 		if !path.IsAbs(requested) {
@@ -484,20 +484,4 @@ func hostConfigFromOptions(options *types.ImageBuildOptions) *container.HostConf
 		ExtraHosts: options.ExtraHosts,
 	}
 	return hc
-}
-
-// fromSlash works like filepath.FromSlash but with a given OS platform field
-func fromSlash(path, platform string) string {
-	if platform == "windows" {
-		return strings.Replace(path, "/", "\\", -1)
-	}
-	return path
-}
-
-// separator returns a OS path separator for the given OS platform
-func separator(platform string) byte {
-	if platform == "windows" {
-		return '\\'
-	}
-	return '/'
 }

